be/src/exprs/agg/window_funnel.h

```diff
@@ -0,0 +1,217 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#pragma once
+
+#include <cstring>
+#include <limits>
+#include <type_traits>
+
+#include "column/array_column.h"
+#include "column/binary_column.h"
+#include "column/datum.h"
+#include "column/fixed_length_column.h"
+#include "column/type_traits.h"
+#include "column/vectorized_fwd.h"
+#include "common/logging.h"
+#include "exprs/agg/aggregate.h"
+#include "gen_cpp/Data_types.h"
+#include "gutil/casts.h"
+#include "runtime/mem_pool.h"
+#include "thrift/protocol/TJSONProtocol.h"
+#include "udf/udf_internal.h"
+#include "util/phmap/phmap_dump.h"
+#include "util/slice.h"
+
+namespace starrocks::vectorized {
+
+struct ComparePairFirst final {
+    template <typename T1, typename T2>
+    bool operator()(const std::pair<T1, T2>& lhs, const std::pair<T1, T2>& rhs) const {
+        return std::tie(lhs) < std::tie(rhs);
+    }
+};
+
+struct WindowFunnelState {
+    // first args is timestamp, second is event position.
+    using TimestampEvent = std::pair<int64_t, uint8_t>;
+    std::vector<TimestampEvent> events_list;
+    int64_t window_size;
+    uint8_t events_size;
+    bool sorted = true;
+
+    void sort() { std::stable_sort(std::begin(events_list), std::end(events_list)); }
+
+    void update(int64_t timestamp, uint8_t event_level) {
+        // keep only one as a placeholder for someone group.
+        if (events_list.size() > 0 && event_level == 0) {
+            return;
+        }
+
+        if (sorted && events_list.size() > 0) {
+            if (events_list.back().first == timestamp)
+                sorted = events_list.back().second <= event_level;
+            else
+                sorted = events_list.back().first <= timestamp;
+        }
+        events_list.emplace_back(std::make_pair(timestamp, event_level));
+    }
+
+    void deserialize_and_merge(DatumArray& datum_array) {
+        if (datum_array.size() == 0) {
+            return;
+        }
+
+        std::vector<TimestampEvent> other_list;
+        window_size = datum_array[0].get_int64();
+        events_size = (uint8_t)datum_array[1].get_int64();
+        bool other_sorted = (uint8_t)datum_array[2].get_int64();
+
+        for (size_t i = 3; i < datum_array.size() - 1; i += 2) {
+            int64_t timestamp = datum_array[i].get_int64();
+            int64_t event_level = datum_array[i + 1].get_int64();
+            other_list.emplace_back(std::make_pair(timestamp, uint8_t(event_level)));
+        }
+
+        const auto size = events_list.size();
+
+        events_list.insert(std::begin(events_list), std::begin(other_list), std::end(other_list));
```

Comment:
```suggestion
        events_list.insert(std::end(events_list), std::begin(other_list), std::end(other_list));
```
events_list before other_list